tests/test_twitter_ads_bookmark.py

```diff
@@ -0,0 +1,182 @@
+                    second_bookmark_value_utc = self.convert_state_to_utc(
+                        second_bookmark_value)
+
+
+                    simulated_bookmark_value = self.convert_state_to_utc(new_states['bookmarks'][stream])
+
+                    # Verify the first sync sets a bookmark of the expected form
+                    self.assertIsNotNone(first_bookmark_value)
+
+                    # Verify the second sync sets a bookmark of the expected form
+                    self.assertIsNotNone(second_bookmark_value)
+
+                    # Verify the second sync bookmark is Equal to the first sync bookmark
+                    # assumes no changes to data during test
+                    self.assertEqual(second_bookmark_value, first_bookmark_value)
+
+                    for record in first_sync_messages:
+
+                        # Verify the first sync bookmark value is the max replication key value for a given stream
+                        replication_key_value = record.get(replication_key)
+
+                        self.assertLessEqual(
+                            replication_key_value, first_bookmark_value_utc,
+                            msg="First sync bookmark was set incorrectly, a record with a greater replication-key value was synced."
+                        )
+
+                    for record in second_sync_messages:
+                        # Verify the second sync replication key value is Greater or Equal to the first sync bookmark
+                        replication_key_value = record.get(replication_key)
+
+                        self.assertGreaterEqual(replication_key_value, simulated_bookmark_value,
+                                                msg="Second sync records do not repect the previous bookmark.")
+
+                        # Verify the second sync bookmark value is the max replication key value for a given stream
+                        self.assertLessEqual(
+                            replication_key_value, second_bookmark_value_utc,
+                            msg="Second sync bookmark was set incorrectly, a record with a greater replication-key value was synced."
+                        )
+
+                elif expected_replication_method == self.FULL_TABLE:
+
+                    # Verify the syncs do not set a bookmark for full table streams
+                    self.assertIsNone(first_bookmark_value)
+                    self.assertIsNone(first_bookmark_value)
+
+                    # Verify the number of records in the second sync is the same as the first
+
+                    # `targeting_criteria` is child stremas of parent stream `line_items` and tickets is incremental streams
+                    # Child stream also behave like incremental streams but does not save it's own state. So, it don't 
+                    # have same no of record on second sync and first sync.
+                    if stream in ['targeting_criteria']:
+                        continue
```

Comment:
Is it possible to add some assertions specific to this child stream? My assumption is that all `target_criteria` records will be replicated that correspond to a given parent object. As the comment says the parent stream is incremental. I think it should be possible to verify consistency of this child stream across the two syncs in this test. On the initial sync we'll have all parent records and all child records for each parent. On the second sync we should have all child records for each parent record with a replication key value greater or equal to the previous bookmark. This should be possible to verify if the parent primary key is a foreign key on the child record.

Reply:
@kspeer825 We have made the child stream an independent stream from the parent stream [PR#18](https://github.com/singer-io/tap-twitter-ads/pull/18). So, we have written a child-specific assertion into it. Actually, during the sync of a child, we are retrieving parents based on the bookmark stored for the child stream. Then, based on the primary key of the parent record, we retrieve the child's records. We are not storing these parent records in the target. So, the scenario that you suggested is difficult to cover here. Instead, we are verifying that the `foreign_keys` list from the second sync must be a subset of the `foreign_keys` list from the first sync.

Reply:
Got it thanks.